98_Drivers/Timers.md

```diff
@@ -0,0 +1,283 @@
+
+This SDT has the standard header, followed by the following fields:
+
+```c
+struct HpetSdt {
+    ACPISDTHeader header;
+    uint32_t event_timer_block_id;
+    uint32_t reserved;
+    uint64_t address;
+    uint8_t id;
+    uint16_t min_ticks;
+    uint8_t page_protection;
+}__attribute__((packed));
+```
+
+*Authors note: the reserved field before the address field is actually some type information describing the address space where the HPET registers are located. In the ACPI table this reserved field is the first part of a 'generic address structure', however we can safely ignore this info because the HPET spec requires the registers to be memory mapped (thus in memory space).*
+
+We're mainly interested in the `address` field which gives us the physical address of the HPET registers. The other fields are explained in the HPET specification but are not needed for our purposes right now.
+
+As with any MMIO you will need to map this physical address into the virtual address space so we can access the registers with paging enabled.
+
+### Theory Of Operation
+
+The HPET consists of a single main counter (that counts up) with some global configuration, and a number of comparators that can trigger interrupts when certain conditions are met in relation to the main counter. The HPET will always have at least 3 comparators, but may have up to 32.
+
+The HPET is similar to the PIT in that we are told the frequency of it's clock. Unlike the PIT, the HPET spec does not give us the frequency directly, we have to read it from the HPET registers. 
+
+Each register is accessed by adding an offset to the base address we obtained before. The main registers we're interested in are:
+
+- General capabilities: offset 0x0.
+- General configuration: offset 0x10.
+- Main counter value: 0xF0.
+
+We can read the main counter at any time, which is measured in in timer ticks. We can convert these ticks into realtime by multiplying them with the timer period in the general capabilities register. Bits 63:32 of the general capabilities register contain the number of femtoseconds for each tick. A nanosecond is 1000 femtoseconds, and 1 second is 1'000'000'000 femtoseconds.
+
+We can also write to the main counter, usually you would write a 0 here when initializing the HPET in order to be able to determine uptime, but this is not really necessary.
+
+The general capabilities register contains some other useful information, like whether the main counter is 32 or 64 bits wide, and the number of comparators (which are used to generate interrupts). These are all detailed in the public spec, and you can explore these at your leisure.
+
+In order for the main counter to actually begin counting, we need to enable it. This is done by setting bit 0 of the general configuration register. Once this bit is set, the main counter will increment by one every time it's internal clock ticks. The period of this clock is what's specified in the general capabilities register (bits 63:32).
+
```

Comment:
could be useful adding a small table with a recap of the hpet registers that we are most likely to use? It would be useful sine someone that is maybe reading the guide, using it as reference while writing the hpet, will have not only the text explaining the bits, but a small table easy to search while writing code? 

what do you think?